ros_ign_gazebo/src/optical_frame_publisher.cpp

```diff
@@ -0,0 +1,200 @@
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2_ros/static_transform_broadcaster.h>
+
+#include <memory>
+#include <string>
+
+#include <geometry_msgs/msg/transform_stamped.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <sensor_msgs/msg/camera_info.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+namespace ros_ign_gazebo
+{
+struct OpticalFramePublisher::Impl
+{
+  /// \brief Check for downstream subscriptions and enable upstream
+  /// subscriptions based on that
+  void CheckSubscribers();
+
+  /// \brief Start upstream subscription for images
+  void ImageConnect();
+
+  /// \brief Start upstream subscription for camera info
+  void CameraInfoConnect();
+
+  /// \brief Image callback
+  void UpdateImageFrame(const sensor_msgs::msg::Image & msg);
+
+  /// \brief Camera info callback
+  void UpdateCameraInfoFrame(const sensor_msgs::msg::CameraInfo & msg);
+
+  /// \brief Publish new TF between frame and optical_frame
+  void PublishTF(const std::string & frame, const std::string & child_frame);
+
+  /// \brief Interface for creating publications/subscriptions
+  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics;
+
+  /// \brief timer with callback to check for publisher subscription count
+  rclcpp::TimerBase::SharedPtr timer;
+
+  /// \brief Static transform broadcaster that broadcasts the optical frame id
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> tfBroadcasterStatic;
+
+  /// \brief Publisher for the image optical frame topic
+  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub;
+
+  /// \brief Publisher for the camera info optical frame topic
+  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr info_pub;
+
+  /// \brief Subscriber for the original image topic
+  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub;
+
+  /// \brief Subscriber for the original camera_info topic
+  rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr info_sub;
+
+  /// \brief Publish camera info or just the image
+  bool publish_camera_info;
+
+  /// \brief New frame id of published messages
+  std::string new_frame_id;
+};
+
+void OpticalFramePublisher::Impl::CheckSubscribers()
+{
+  if (image_pub->get_subscription_count() > 0u) {
+    ImageConnect();
+  }
+
+  if (info_pub && info_pub->get_subscription_count() > 0u) {
+    CameraInfoConnect();
+  }
+}
+
+void OpticalFramePublisher::Impl::ImageConnect()
+{
+  // Skip connecting if publisher is already created (non-nullptr)
+  if (image_pub) {
+    return;
+  }
+
+  image_sub = rclcpp::create_subscription < sensor_msgs::msg::Image > (
+    node_topics,
+    "input/image", 10,
+    std::bind(&OpticalFramePublisher::Impl::UpdateImageFrame, this, _1));
+}
+
+void OpticalFramePublisher::Impl::CameraInfoConnect()
+{
+  // Skip connecting if publisher is already created (non-nullptr)
+  if (info_pub) {
```

Comment:
Shouldn't this be `info_sub`? I think we should skip if we already created the subscriber, right?